In [1]:
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns



In [5]:
# Loading the Dataset

df = pd.read_csv("SzegedHungaryWeatherHistory.csv")

# Previewing the DataSet

print(df.head())
print(df.info())


                  Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                     251          15.8263           0               1015.13  

In [8]:
# Checking for the missing values in our dataset

print(df.isnull().sum())

# Checking the connection by selecing only the numeric columns from the dataset for this

numeric_df = df.select_dtypes(include=['number'])
print(numeric_df.corr()['Temperature (C)'])

Formatted Date                0
Summary                       0
Precip Type                 517
Temperature (C)               0
Apparent Temperature (C)      0
Humidity                      0
Wind Speed (km/h)             0
Wind Bearing (degrees)        0
Visibility (km)               0
Loud Cover                    0
Pressure (millibars)          0
Daily Summary                 0
dtype: int64
Temperature (C)             1.000000
Apparent Temperature (C)    0.992629
Humidity                   -0.632255
Wind Speed (km/h)           0.008957
Wind Bearing (degrees)      0.029988
Visibility (km)             0.392847
Loud Cover                       NaN
Pressure (millibars)       -0.005447
Name: Temperature (C), dtype: float64


In [ ]:
# Dropping the missing values from the dataset

df = df.dropna()

# Defining multiple features (X) and target (y) variables

X = df[['Humidity', 'Wind Speed (km/h)', 'Pressure (millibars)']]
y = df['Temperature (C)']        

# Splitting the dataset into training and testing sets

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)           


In [12]:
from sklearn.linear_model import LinearRegression

# Create & Train the model

model = LinearRegression()
model.fit(X_train, y_train)

# Making the predictions

y_pred = model.predict(X_test)

In [13]:
from sklearn.metrics import mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"New Mean Squared Error: {mse:.2}")
print(f"New R-squared Score: {r2:.2}")

New Mean Squared Error: 5.3e+01
New R-squared Score: 0.42


## CONCLUSION For This New "Multiple Linear Regression" Model:

1. **Model Evolution:** I transitioned from Simple Linear Regression (1 feature) to Multiple Linear Regression (3 features: Humidity, Wind Speed, and Pressure).
2. **Performance Boost:** The R-squared score increased from 0.26 to 0.42, meaning the model now explains 42% of temperature variations.
3. **Error Reduction:** The Mean Squared Error dropped significantly from 144.08 to 53.0. This indicates that adding more weather variables makes the predictions much more reliable.
4. **Key Learning:** Temperature is a complex variable. While humidity is a strong predictor, combining it with wind speed and atmospheric pressure provides a much more accurate "picture" for the machine learning model.